gateway/src/test/java/io/camunda/zeebe/gateway/api/job/StreamActivatedJobsTest.java

```diff
@@ -0,0 +1,182 @@
+    jobRecord.setType(jobType);
+    final Map<String, Object> fetchedVariables = Map.of("foo", 1, "bar", 2);
+    jobRecord.setVariables(MsgPackUtil.asMsgPack(fetchedVariables));
+    activatedJob.setRecord(jobRecord);
+
+    jobStreamer.push(activatedJob).join();
+
+    // then
+    final ActivatedJob streamedActivatedJob = streamedJobs.get(0);
+    assertThat(streamedActivatedJob.getType()).isEqualTo(activatedJob.jobRecord().getType());
+    assertThat(activatedJob.jobRecord().getVariables()).isEqualTo(fetchedVariables);
+  }
+
+  @Test
+  public void shouldPushNoJobsWhenNonAvailable() {
+    // given
+    final String jobType1 = "testJob1";
+    final String jobType2 = "testJob2";
+    final String worker = "testWorker";
+    final Duration timeout = Duration.ofMinutes(1);
+    final List<String> fetchVariables = List.of("foo");
+
+    final List<ActivatedJob> streamedJobs1 =
+        getStreamActivatedJobsRequest(jobType1, worker, timeout, fetchVariables).streamedJobs;
+    getStreamActivatedJobsRequest(jobType2, worker, timeout, fetchVariables);
+
+    // when
+    final ActivatedJobImpl activatedJob = new ActivatedJobImpl();
+    final JobRecord jobRecord = new JobRecord();
+    jobRecord.setType(jobType2); // only push job for job type 2
+    activatedJob.setRecord(jobRecord);
+
+    jobStreamer.push(activatedJob).join();
+
+    // then
+    assertThat(streamedJobs1).isEmpty();
+  }
+
+  @Test
+  public void shouldRemoveStreamOnError() {
+    // given
+    final String jobType = "testJobOnError";
+    final String worker = "testWorker";
+    final Duration timeout = Duration.ofMinutes(1);
+    final List<String> fetchVariables = List.of("foo");
+
+    final TestStreamObserver streamObserver =
+        getStreamActivatedJobsRequest(jobType, worker, timeout, fetchVariables);
+
+    assertThat(jobStreamer.containsStreamFor(jobType)).isTrue();
+
+    // when
+    streamObserver.onError(new RuntimeException("test on error"));
+
+    // then
+    Awaitility.await("until the stream is removed")
+        .until(() -> !jobStreamer.containsStreamFor(jobType));
+  }
+
+  @Test
+  public void shouldRemoveStreamOnCancel() {
+    // given
+    final String jobType = "testJobOnCancel";
+    final String worker = "testWorker";
+    final Duration timeout = Duration.ofMinutes(1);
+    final List<String> fetchVariables = List.of("foo");
+    final TestStreamObserver streamObserver =
+        getStreamActivatedJobsRequest(jobType, worker, timeout, fetchVariables);
+    assertThat(jobStreamer.containsStreamFor(jobType)).isTrue();
+
+    // when
+    streamObserver.cancel();
+
+    // then
+    Awaitility.await("until the stream is removed")
+        .until(() -> !jobStreamer.containsStreamFor(jobType));
+  }
+
+  // TODO add test for onClose
+  // How to test onClose? it seems the graceful way to close a stream from the client side is
+  // simply to cancel it, as per the gRPC docs. So is onClose only for when we close it server side?
```

Comment:
:thinking: From my understanding, `onClose` is called when the stream is completed gracefully or with an error. So you already test that with the `shouldRemoveStreamOnError()` test case. 

You can also add `requestStream.onCompleted();` in the `TestStreamObserver#onCompleted()` method, and add an additional test case for completing the stream with the `streamObserver` instance.

//cc @npepinpe are my assumptions correct?

Reply:
@koevskinikola Current test cases only trigger `onCancelHandler` hook. It is still not clear how we can trigger `onCloseHandler`. I think, it is better if we create an issue for that and have a separate PR (as we did for validations). **What do yo think?**

AFAIK, there isn't a way to set hooks for handling `onCompleted`. Based on the design, we need to remove the worker from the stream. The only way to de-register streams is to cancel the RPC call. Therefore, we decided not to have a case where client calls `onCompleted`. I think, client implementation will apply the same design principle.

Cc: @npepinpe

Reply:
I guess we should cover this case in the IT tests then. It would be good if you note this in the issue you create for them.